docs/design_proposals/digest-tagger.md

```diff
@@ -0,0 +1,81 @@
+
+The default tagger, if none is specified in the `skaffold.yaml`, is the `git` tagger.
+
+Here are some rules about how tagging currently works:
+
+ + Image tags are computed before the images are built.
+ + In early versions of skaffold, we tried to compute tags after the build. It made the process
+   super complex with a lot of retagging. It also produced images that were tagged with their
+   own digest or imageID which is superfluous since the digest/imageID can be used to reference
+   the images directly.
+ + No matter the tagger, Skaffold always uses immutable referenced in Kubenetes manifests.
+   Which reference is used depend on whether the images are pushed or not:
+     + When images are pushed, their immutable digests are available. So images are referenced
+       by image tag and digest. Something like `image:tag@sha256:abacabac...`. Using both the
+       tag and the digest seems superfluous but it guarantees that Kubernetes uses the exact image
+       version, referenced by its immutable digest while helping the user quickly see which
+       version of the image is used.
+     + When images are NOT pushed, digests are not available. All we have is the tags
+       and imageIDs. Since imageIDs can't be used in Kubernetes manifests, Skaffold creates
+       an additional tag with the same name as the imageID and uses that in manifests.
+       This ensures that the exact image version is used.
+ + We can never reference images just by their tags because those tags are mutable and
+   can lead to cases where Kubernetes will use an outdated version of the image.
+
+## Issues
+
+ + the `git` tagger requires users to have the `git` binary installed. It also requires
+   the project to be a git project, which is typically not the case when users just try
+   to get started.
+ + the `sha256` has a wrong name. It is named like that because, in the end, when Skaffold
+   deploys to a remote cluster, the image's sha256 digest will be used as the immutable tag.
+   Users are very confused with this name and behaviour.
+ + the `sha256` used to be able to use the image tags provided in the artifact definition,
+   instead of `latest`. This was not documented and is not possible anymore because artifact
+   definitions are now considered invalid if images names have a tag.
+ + the `envTemplate` tagger used to be able to replace `{{.DIGEST}}` with the image's imageID
+   or digest. This was buggy and is not possible anymore since tags are computed before the
+   images are built.
+ + users want a tagger that uses the inputs' digest as a tag. They think that's what the
+   `sha256` tagger should do.
+   
+## Proposal
+
+ + we introduce a `latest` tagger tag tags images with `:latest`.
+ + the `latest` tagger is used by default instead of the `git` tagger.
+ + `sha256` tagger is deprecated.
+ + An `inputDigest` is added. It uses the digest of the artifact's inputs as the tag.
```

Comment:
Isnt this duplicating what docker does ?

Reply:
except skaffold does it before the build and docker does it during build.

Reply:
there is a big difference: docker calculates image content digests we calculate input source digests. The two can differ in case there are external dependencies / sources of randomness or other reasons the build is not deterministic.

Reply:
ok.